Update testing section of porting guide (#53)
MCU-Driver-HAL/MCU-Driver-HAL: docs/porting/porting_full_process/built_in_tests.md

# Testing your port

It's important to test your port at the end of each module porting, rather than only once when you've imported all modules. There are two testing methods:

1. Running the Mbed OS built-in tests with *Greentea*.

> 1. Manually running the Mbed OS built-in tests.

Reviewer: Are these statements still true (minus the Mbed OS part)?

Author: No, I believe we don't have semi-automatic testing in MCU-Driver-HAL yet (e.g. running `mbed test` which would build and run all greentea tests for the connected targets).

Reviewer 2: We have semi-automatic tests although I'm not fuzzy about this terminology. The instructions are ok IMO.

### Prerequisites

#### Minimum HAL module support

To run the built-in tests, you need to have ported and verified at least these HAL modules:

> - Low power ticker.

Reviewer 2: I don't think we need LPticker. We consider usTicker sufficient.
```suggestion
```

# Testing your port

It's important to test your port at the end of each module porting, rather than only once when you've imported all modules. 

## Testing with the Greentea framework

### Prerequisites

#### Minimum HAL module support

To run the built-in tests, you need to have ported and verified at least these HAL modules:

- Serial port (synchronous transfer). To verify that it works, load a test binary with `printf()`. Verify debug messages can be seen on your serial terminal program.
- Microsecond ticker.

You'll also need to have ported DAPLink or compatible interface firmware to your interface chip.
    <span class="notes">If DAPLink is still under development, you can still run tests manually.</span>

#### mbed-tools

Some tools from Mbed, including `mbedhtrun`, are required to run Greentea tests.

See [docs](../../user/README.md) about tools setup.

> ### Compiling tests

Reviewer 3: This section and the following (running the tests) are platform specific (even if they share a common theme).
That IMHO do not belong here.

The expected output, to give a rough idea of what to expect when running the tests, is fine but the first line  (with the command) needs to be removed.

Reviewer 2: I assume you mean `Example command line` in lines 46-49. 
I tend to agree and prefer this to be removed. There is another example below.

Reviewer 3: I mean moving lines  25-57 to the relevant SiP repo and removing line 63-64 without necessarily moving them to somewhere else.

Author: Aside from the paths in the examples, the rest isn't platform specific, no? I we remove everything regarding compiling and running tests there won't be much point having this file/guide at all.

Reviewer 2: See suggestions on minor things to tweak.

> 1. Go to the directory containing the `CMakeLists.txt` file of the test you want to run, e.g. `MCU-Driver-Arm/sdfx_arm/tests/mbed_hal/echo/`.

Reviewer 2: ```suggestion
1. Go to the directory containing the `CMakeLists.txt` file of the test you want to run.
```

2. Generate the build system files
   ```
   cmake -S . -B cmake_build
   ```

>    The default toolchain is GCC_ARM. You can use a different toolchain by passing `-D MBED_TOOLCHAIN=<TOOLCHAIN>`. If you need to debug the binary, build with `-D CMAKE_BUILD_TYPE=debug`.

Reviewer 2: ```suggestion
   The default toolchain is GCC_ARM. You can use a different toolchain by passing `-D MBED_TOOLCHAIN=<TOOLCHAIN>`. The default CMake build type option is `develop`, best for debugging.
```

4. Compile the test
   ```
   cmake --build cmake_build -j
   ```

### Running the tests

Use `mbedhtrun` with the following options to run a test:
* `-f <path/to/test/binary.bin>` This is the path to the binary built above.
* `-e <path/to/host_tests/directory>` This will usually be `/MCU-Driver-HAL/tests/host_tests`.
* `-d <path/to/DAPLINK/mount/point>` This is the place where you usually drag-and-drop binaries when you flash your hardware via DAPLINK.
* `-p <path/to/tty/port>:115200` This is the serial interface that Greentea will use to talk to your hardware.

Example command line:
```

> mbedhtrun -f ./sdfx_st/tests/mbed_hal/echo/cmake_build/sdfx-st-test-echo.bin -e ./MCU-Driver-HAL/tests/host_tests -d /Volumes/DIS_L4IOT -p /dev/tty.usbmodem145103:115200

Reviewer 2: ```suggestion
mbedhtrun -f ./tests/mbed_hal/echo/cmake_build/....bin -e ./MCU-Driver-HAL/tests/host_tests -d /Volumes/XXX -p /dev/tty.usbmodem145103:115200
```

```

If your hardware doesn't have a DAPLINK interface, you must flash the test binary manually via other means, then manually reset the hardware. Then, run `mbedhtrun` with the `--skip-reset --skip-flashing` options, omitting the `-f` and `-d` arguments.

Example command line (without DAPLINK):
```

> mbedhtrun -e ./MCU-Driver-HAL/tests/host_tests -p /dev/tty.usbmodem145103:115200 --skip-reset --skip-flashing

Reviewer 2: ```suggestion
mbedhtrun -e ./MCU-Driver-HAL/tests/host_tests -p /dev/ttyXXX:115200 --skip-reset --skip-flashing
```

```

#### Expected output

Greentea output for a test that ran successfully looks like this:

```

> mbedhtrun -f ./sdfx_st/tests/mbed_hal/echo/cmake_build/sdfx-st-test-echo.bin -e ./MCU-Driver-HAL/tests/host_tests -d /Volumes/DIS_L4IOT -p /dev/tty.usbmodem145103:115200

Reviewer 2: ```suggestion
mbedhtrun -f ./tests/mbed_hal/echo/cmake_build/s.....bin -e ./MCU-Driver-HAL/tests/host_tests -d /Volumes/XXX -p /dev/tty.usbmodem145103:115200
```

You'll also need to have ported DAPLink or compatible interface firmware to your interface chip.
    <span class="notes">If DAPLink is still under development, you can still run tests manually.</span>

#### mbed-tools

Some tools from Mbed, including `mbedhtrun`, are required to run Greentea tests.

Run `pip install mbed-tools` to install the required tools.

### Compiling tests

> 1. Go into the directory of the test for the module you want to test, e.g. `MCU-Driver-Arm/sdfx_arm/tests/mbed_hal/echo/`.

Reviewer: ```suggestion
1. Go to the directory containing the CMake build instructions for the test application binary you would like to run.
```

Author: Made some changes to this

> 2. Run `cmake -S . -B cmake_build` to generate the build system files.

Reviewer: ```suggestion
1. Generate the build system files.
    ````
    cmake -S . -B cmake_build -D GREENTEA_CLIENT_STDIO=OFF -D MBED_TOOLCHAIN=<TOOLCHAIN> -D CMAKE_BUILD_TYPE=debug
    ````
```

Author: MBED_TOOLCHAIN doesn't seem to be needed.

Reviewer 2: It's needed, according to the build scripts
https://github.com/MCU-Driver-HAL/MCU-Driver-ST/blob/main/sdfx_st/tests/build_all.sh#L40

Author: If not provided, it defaults to GCC_ARM here https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/blob/efab67e6bbe6120d2e0418f16698f21bbd1782dd/CMakeLists.txt#L18

You only need to pass it if you want to use a different toolchain.

Also regarding GREENTEA_CLIENT_STDIO, I've had another look and I believe it's not needed. It's OFF by default anyways and it seems we're always using greentea-custom_io in MCU-Driver-HAL so setting GREENTEA_CLIENT_STDIO=ON would just break all our tests.

Reviewer 2: I like simpler instructions and I'm glad to hear that we can get rid of `GREENTEA_CLIENT_STDIO`.

We need to support two toolchains (GCC_ARM and ARM) but I see you already added instructions on how users can do it.

> 3. Run `cmake --build cmake_build -j` to compile the test.

Reviewer: ```suggestion
3. Compile the test.
    ````
    cmake --build cmake_build -j
    ````
```

Author: Applied

### Running the tests

> You can use `mbedhtrun` with the following options to run a test:

Reviewer: ```suggestion
Use `mbedhtrun` with the following options to run a test:
```

* `-f <path/to/test/binary.bin>` This is the path to the binary built above.
* `-e <path/to/host_tests/directory>` This will usually be `/MCU-Driver-HAL/tests/host_tests`.
* `-d <path/to/DAPLINK/mount/point>` This is the place where you usually drag-and-drop binaries when you flash your hardware via DAPLINK.
* `-p <path/to/tty/port>:115200` This is the serial interface that Greentea will use to talk to your hardware.

> Example command line: `mbedhtrun -f ./sdfx_st/tests/mbed_hal/echo/cmake_build/sdfx-st-test-echo.bin -e ./MCU-Driver-HAL/tests/host_tests -d /Volumes/DIS_L4IOT -p /dev/tty.usbmodem145103:115200`

Reviewer: ```suggestion
e.g
 ``
  mbedhtrun -f ./sdfx_st/tests/mbed_hal/echo/cmake_build/sdfx-st-test-echo.bin -e ./MCU-Driver-HAL/tests/host_tests -d /Volumes/DIS_L4IOT -p /dev/tty.usbmodem145103:115200
 ``
```

Reviewer 2: pending to review/update

Author: I think it's ok as is

> If your hardware doesn't have a DAPLINK interface, you can flash the test binary manually via other means, manually reset the hardware, omit the `-d <path/to/DAPLINK/mount/point>` option and pass the `--skip-reset --skip-flashing` options to `mbedhtrun`.

Reviewer: ```suggestion
If your hardware does not have a DAPLINK interface, you can program the test app binary via other means. After manually resetting the hardware, omit the `-d` and `-f` optional command line arguments and pass the `--skip-reset --skip-flashing` argument to `mbedhtrun`.
```

Reviewer 2: pending to review/update

Author: I made some changes to this part, I think it looks ok now.

> Example command line (without DAPLINK): `mbedhtrun -f ./sdfx_st/tests/mbed_hal/echo/cmake_build/sdfx-st-test-echo.bin -e ./MCU-Driver-HAL/tests/host_tests -p /dev/tty.usbmodem145103:115200 --skip-reset --skip-flashing`

Reviewer: ```suggestion
e.g
``
mbedhtrun -f ./sdfx_st/tests/mbed_hal/echo/cmake_build/sdfx-st-test-echo.bin -e ./MCU-Driver-HAL/tests/host_tests -p /dev/tty.usbmodem145103:115200 --skip-reset --skip-flashing
``
```

Reviewer 2: pending to review/update

Author: I think it's ok as is

#### Expected output

> Greetea output for a test that ran successfully looks like this:

Reviewer: ```suggestion
The Greentea output for a test that ran successfully would be similar to the following:
```

Reviewer 2: pending to review/update

Author: I think it's ok as is

[1626879930.01][HTST][INF] CONN exited with code: 0
[1626879930.01][HTST][INF] Some events in queue
[1626879930.01][HTST][INF] stopped consuming events
[1626879930.01][HTST][INF] host test result() call skipped, received: True
[1626879930.01][HTST][INF] calling blocking teardown()
[1626879930.01][HTST][INF] teardown() finished
[1626879930.01][HTST][INF] {{result;success}}
```

## Manual testing

You may want to run tests manually, for example if DAPLink is still under development. You will need to export your tests from Greentea and import them to your IDE. For example:

1. Find the test directory:

    ```
    mbed test -m <new_target> -t gcc_arm --compile-list -n mbed-os-tests-mbed_hal-common_ticker
    ```

1. Copy the source code to the project root directory:

    ```
    cd <separate folder from existing porting project>
    cp -R mbed-os-example-blinky/mbed-os/TESTS/mbed_hal/common_tickers .
    cd common_tickers
    mbed new --create-only .
    ```

1. Copy the `mbed-os` directory:

    ```
    cp -R ../mbed-os-example-blinky/mbed-os .
    ```

1. Export to a makefile project:

    ```
    mbed export -i <exporter> -m <new_target>
    ```

> 1. Flash the program with pyOCD (using the same configuration you used when you initially set up pyOCD).

Reviewer: This might still be possible.

Reviewer 2: Although it's possible, it's very specific. htrun handles the flash through mass storage.
I'm ok to remove this.

You'll also need to have ported DAPLink or compatible interface firmware to your interface chip.
    <span class="notes">If DAPLink is still under development, you can still run tests manually.</span>

#### mbed-tools

Some tools from Mbed, including `mbedhtrun`, are required to run Greentea tests.

> Run `pip install mbed-tools` to install the required tools.
> 

Reviewer 2: ```suggestion
See [docs](../../docs/user/README.md) about tools setup.
```

### Running the tests

Use `mbedhtrun` with the following options to run a test:
* `-f <path/to/test/binary.bin>` This is the path to the binary built above.
* `-e <path/to/host_tests/directory>` This will usually be `/MCU-Driver-HAL/tests/host_tests`.
* `-d <path/to/DAPLINK/mount/point>` This is the place where you usually drag-and-drop binaries when you flash your hardware via DAPLINK.
* `-p <path/to/tty/port>:115200` This is the serial interface that Greentea will use to talk to your hardware.

Example command line: `mbedhtrun -f ./sdfx_st/tests/mbed_hal/echo/cmake_build/sdfx-st-test-echo.bin -e ./MCU-Driver-HAL/tests/host_tests -d /Volumes/DIS_L4IOT -p /dev/tty.usbmodem145103:115200`

If your hardware doesn't have a DAPLINK interface, you must flash the test binary manually via other means, then manually reset the hardware. Then, run `mbedhtrun` with the `--skip-reset --skip-flashing` options, omitting the `-f` and `-d` arguments.

Example command line (without DAPLINK): `mbedhtrun -e ./MCU-Driver-HAL/tests/host_tests -p /dev/tty.usbmodem145103:115200 --skip-reset --skip-flashing`

#### Expected output

Greetea output for a test that ran successfully looks like this:

```

> MCU-Driver-ST % mbedhtrun -f ./sdfx_st/tests/mbed_hal/echo/cmake_build/sdfx-st-test-echo.bin -e ./MCU-Driver-HAL/tests/host_tests -d /Volumes/DIS_L4IOT -p /dev/tty.usbmodem145103:115200

Reviewer 2: Remove details on specific environment
```suggestion
mbedhtrun -f ./sdfx_st/tests/mbed_hal/echo/cmake_build/sdfx-st-test-echo.bin -e ./MCU-Driver-HAL/tests/host_tests -d /Volumes/DIS_L4IOT -p /dev/tty.usbmodem145103:115200
```
